blocks/math/include/gnuradio-4.0/math/ExpressionBlocks.hpp

```diff
@@ -0,0 +1,352 @@
+#ifndef EXPRESSIONBLOCKS_HPP
+#define EXPRESSIONBLOCKS_HPP
+
+#include <algorithm>
+#include <gnuradio-4.0/Block.hpp>
+#include <gnuradio-4.0/BlockRegistry.hpp>
+
+#include <exprtk.hpp>
+
+namespace gr::blocks::math {
+
+namespace detail {
+
+inline std::string formatParserError(const auto& parser, std::string_view expression) {
+    std::stringstream ss;
+    for (std::size_t i = 0; i < parser.error_count(); ++i) {
+        const auto error = parser.get_error(i);
+
+        ss << fmt::format("ExprTk Parser Error({:2}):  Position: {:2}\nType: [{:14}] Msg: {}; expression:\n{}\n", //
+            static_cast<unsigned int>(i),                                                                         //
+            static_cast<unsigned int>(error.token.position),                                                      //
+            exprtk::parser_error::to_str(error.mode), error.diagnostic, expression);
+    }
+    return ss.str();
+}
+
+struct VectorInfo {
+    std::size_t size;
+    ssize_t     index;
+};
+
+inline VectorInfo computeVectorInfo(void* base_ptr, void* end_ptr, std::size_t elementSize, void* access_ptr) {
+    if (!base_ptr || !end_ptr || !access_ptr) {
+        throw std::invalid_argument("null pointer(s) provided.");
+    }
+
+    auto base   = static_cast<std::byte*>(base_ptr);
+    auto end    = static_cast<std::byte*>(end_ptr);
+    auto access = static_cast<std::byte*>(access_ptr);
+
+    if (end < base) {
+        throw std::out_of_range(fmt::format("invalid vector boundaries [{}, {}]", base_ptr, end_ptr));
+    }
+
+    return {static_cast<std::size_t>(end - base) / elementSize, (access - base) / static_cast<ssize_t>(elementSize)};
+}
+
+struct vector_access_rtc : public exprtk::vector_access_runtime_check {
+    std::unordered_map<void*, std::string> vector_map;
+
+    bool handle_runtime_violation(violation_context& context) override {
+        auto               itr         = vector_map.find(static_cast<void*>(context.base_ptr));
+        const std::string& vector_name = (itr != vector_map.end()) ? itr->second : "Unknown";
+
+        const auto typeSize      = static_cast<std::size_t>(context.type_size);
+        auto [vecSize, vecIndex] = computeVectorInfo(context.base_ptr, context.end_ptr, typeSize, context.access_ptr);
+        throw gr::exception(fmt::format("vector access '{name}[{index}]' outside of [0, {size}[ (typesize: {typesize})", //
+            fmt::arg("name", vector_name), fmt::arg("size", vecSize), fmt::arg("index", vecIndex), fmt::arg("typesize", typeSize)));
+        return false; // should never reach here
```

Comment:
Dead code due to the throw directly above.
For custom state handling logic, I see the argument of having such a "never reached" statement, but here we effectively guard against something the compiler guarantees.

Reply:
Unless exceptions are disabled...